### ISSUE TITLE
[YTMusicBackend-3] mopidy_ytmusic translate_uri error "'NoneType' object has no attribute 'get_signature'"

### ISSUE THREAD

**Issue by karl0ss** [OzymandiasTheGreat/mopidy-ytmusic#41]
Guessing the project I have been working on is now dead, as no one seems to respond here, and as of this morning I noticed that the plugin has started having issues any wont play what appears to be any premium music, when i play tracks from MyLikes, it will play stuff I have uploaded and at will play free music i liked from youtube, but will skip any music that is coming from Youtube Music premium, wont play any tracks from albums or anything like that, the error being reported it 

`Nov 22 16:24:16 diskPlayer mopidy[1942]: ERROR    [YTMusicBackend-3] mopidy_ytmusic translate_uri error "'NoneType' object has no attribute 'get_signature'"
Nov 22 16:24:16 diskPlayer mopidy[1942]: WARNING  [Core-10] mopidy.core.tracklist Track is not playable: ytmusic:track:aBhulso8di4`

Like i say, the plugin has been working perfectly, i have run 

`sudo mopidyctl ytmusic reauth`

And it updated my auth.json file with new details from when I set this up, rebooted and stuff, but no difference.

Really hope someone can help/fix

**Comment by impliedchaos** [comment]
Thanks I'll take a look.

**Comment by impliedchaos** [comment]
So I checked and this is actually a bug with PyTube which is what we use to break the signature obfuscation.  https://github.com/pytube/pytube/issues/1163

They've already got pull requests submitted to fix this, so I'm just going to wait for them to make a new release, then I'll update our code to use the new version, and publish a new release as well.  Probably won't take long.

**Comment by karl0ss** [comment]
@impliedchaos just to let you know, things seeeem to be working again today....

**Comment by karl0ss** [comment]
@impliedchaos just to let you know, things seeeem to be working again today...

**Comment by impliedchaos** [comment]
Thanks, that's good to know.  Google changes the javascript obfuscation function daily, and it seems like today's doesn't have the code that caused the previous breakage.  So I'm still waiting on pytube to implement the fix.  Hopefully soon so I can roll out the new release and keep this one from happening again.

**Comment by karl0ss** [comment]
Ok, thanks for the update, I will let you know if anything changes again here in this thread.

tahnks for your great work.

**Comment by zonywhoop** [comment]
FYI pytube version 11.0.2 is now out and appears to fix this issue along with an issue that was preventing me from playing any music at all. I would just receive the below error:

> Dec 22 18:58:32 daplex mopidy[326566]: WARNING  [Core-11] mopidy.core.tracklist Track is not playable: ytmusic:track:5LScCxHS8k8
> Dec 22 18:58:32 daplex mopidy[326566]: ERROR    [YTMusicBackend-3] mopidy_ytmusic translate_uri error "'NoneType' object has no attribute 'get_signature'"

I ran the following to fix upgrade:
`pip3 install -U pytube`

**Comment by impliedchaos** [comment]
Thanks!  I'll update to require this latest pytube version and push out a new release.

**Comment by simu** [comment]
I'm getting `NoneType object has not attribute 'get_signature'` errors again with the current obfuscation (I personally noticed songs no longer playing starting today, 2022-01-20, approx 12:00 UTC). I found https://github.com/pytube/pytube/issues/1199 which suggests a small change in pytube.

Manually applying the suggested change in the regex (https://github.com/pytube/pytube/issues/1199#issuecomment-1016783092) in pytube's `cipher.py` resolved the issue for me for now with pytube-11.0.2 and mopidy-ytmusic 0.3.3

**Comment by sprnza** [comment]
I've applied the suggested change and it had been working for some time but now it's broken again

**Comment by sprnza** [comment]
I've also applied a change mentioned here[here](https://github.com/pytube/pytube/issues/1218#issuecomment-1026499476). The error is gone, but playlist is not being populated

**Comment by sprnza** [comment]
[This is](https://github.com/pytube/pytube/pull/1222) the correct fix I suppose, but it still doesn't work with it.

**Comment by shameerariff** [comment]
Hi there, I am unable to install pytube@nfunc_regex_patch
WARNING: Discarding git+https://github.com/glubsy/pytube@nfunc_regex_patch. Command errored out with exit status 1: git checkout -q nfunc_regex_patch Check the logs for full command output
ERROR: Command errored out with exit status 1: git checkout -q nfunc_regex_patch Check the logs for full command output.
I even tried upgraded the pytube to the latest, still I am having the same problem.

**Comment by Nick1296** [comment]
Take a look at this issue: https://github.com/pytube/pytube/issues/1289 
it solved the problem for me

**Comment by ForumPlayer** [comment]
> Take a look at this issue: [pytube/pytube#1289](https://github.com/pytube/pytube/issues/1289) it solved the problem for me

worked here too

**Comment by shameerariff** [comment]
> Take a look at this issue: [pytube/pytube#1289](https://github.com/pytube/pytube/issues/1289) it solved the problem for me

Worked

**Comment by markgaze** [comment]
Doing the following fixed it for me 🙂 

```
sudo pip uninstall pytube
sudo pip install git+https://github.com/pytube/pytube@master
```

(because the branch that was originally used in @sprnza's has now been merged)

**Comment by ForumPlayer** [comment]
> Take a look at this issue: [pytube/pytube#1289](https://github.com/pytube/pytube/issues/1289) it solved the problem for me

This worked for some time, but broke again yesterday :P

**Comment by rebernet** [comment]
> > Take a look at this issue: [pytube/pytube#1289](https://github.com/pytube/pytube/issues/1289) it solved the problem for me
> 
> This worked for some time, but broke again yesterday :P

same here.

**Comment by OzymandiasTheGreat** [comment]
Here's the new fix https://github.com/pytube/pytube/issues/1326#issuecomment-1144399816

**Comment by mgoltzsche** [comment]
Same problem here with pytube 12.1.3.
When I use the latest pytube version 15.0.0 directly (the CLI), the problem does not appear and I can download tracks.
However, mopidy-ytmusic still requires pytube<13 and therefore gets disabled when mopidy starts and the requirement is not met.

I wonder if this is due to my environment since other issues about other problems were created later within this repo, indicating other people don't have the pytube incompatibility problem somehow, do they?

**Comment by mgoltzsche** [comment]
So how does this work for anyone? I even [patched mopidy-ytmusic](https://github.com/mgoltzsche/mopidy-ytmusic/commit/c60055bc4cbc35534ef4c141fc883928bf5ca280) to use pytube 15.0.0 but it still fails, logging the error:
```
ERROR    2023-09-02 18:11:39,850 [31:YTMusicBackend-11 (_actor_loop)] mopidy_ytmusic
  translate_uri error "'NoneType' object has no attribute 'get_signature'"
WARNING  2023-09-02 18:11:39,850 [31:Core-14 (_actor_loop)] mopidy.core.tracklist
  Track is not playable: ytmusic:track:Vft7l0L8PpM
```

**Comment by cma84** [comment]
> So how does this work for anyone? I even [patched mopidy-ytmusic](https://github.com/mgoltzsche/mopidy-ytmusic/commit/c60055bc4cbc35534ef4c141fc883928bf5ca280) to use pytube 15.0.0 but it still fails, logging the error:
> 
> ```
> ERROR    2023-09-02 18:11:39,850 [31:YTMusicBackend-11 (_actor_loop)] mopidy_ytmusic
>   translate_uri error "'NoneType' object has no attribute 'get_signature'"
> WARNING  2023-09-02 18:11:39,850 [31:Core-14 (_actor_loop)] mopidy.core.tracklist
>   Track is not playable: ytmusic:track:Vft7l0L8PpM
> ```

Did you solve the problem?

Thanks

**Comment by mgoltzsche** [comment]
No, I didn't try it again since then. Instead, I used the [mopidy-youtube](https://github.com/natumbri/mopidy-youtube) plugin (and then switched to [mopidy-spotify](https://github.com/mopidy/mopidy-spotify) recently).

TL;DR:
The mopidy-youtube plugin is better than mopidy-ytmusic anyway when it comes to playing single songs and auto-play, also those from regular YouTube, not just Youtube Music. Also, since it uses the Youtube Data API and the well maintained [yt-dlp](https://github.com/yt-dlp/yt-dlp) library to download audio, mopidy-youtube is more reliable than mopidy-ytmusic which relies on [pytube](https://github.com/pytube/pytube) which doesn't seem to be maintained anymore and implements very brittle HTML parsing logic that breaks regularly. However, as opposed to mopidy-ytmusic mopidy-youtube doesn't list artists and albums in the search view as nicely as the mopidy-ytmusic plugin and correspondingly doesn't make artists and albums nicely navigable - these are the only features I miss from mopidy-ytmusic.

**Comment by cma84** [comment]
> No, I didn't try it again since then. Instead, I used the [mopidy-youtube](https://github.com/natumbri/mopidy-youtube) plugin (and then switched to [mopidy-spotify](https://github.com/mopidy/mopidy-spotify) recently).
> 
> TL;DR: The mopidy-youtube plugin is better than mopidy-ytmusic anyway when it comes to playing single songs and auto-play, also those from regular YouTube, not just Youtube Music. Also, since it uses the Youtube Data API and the well maintained [yt-dlp](https://github.com/yt-dlp/yt-dlp) library to download audio, mopidy-youtube is more reliable than mopidy-ytmusic which relies on [pytube](https://github.com/pytube/pytube) which doesn't seem to be maintained anymore and implements very brittle HTML parsing logic that breaks regularly. However, as opposed to mopidy-ytmusic mopidy-youtube doesn't list artists and albums in the search view as nicely as the mopidy-ytmusic plugin and correspondingly doesn't make artists and albums nicely navigable - these are the only features I miss from mopidy-ytmusic.

Thank you very much for replying, it seems that the YTMusic plugin is not worth the effort.

I have also tried to configure the Mopidy-Youtube plugin but something keeps failing me, as the behavior is similar, I find content in Iris but then it doesn't play. I'm using ChatGPT to try to solve the problem, I've made tests by command line and it downloads the content but in the Iris UI I can't get it to play. I will ask in the add-on forum. 

Translated with DeepL.com (free version)

**Comment by mgoltzsche** [comment]
@cma84 I just tried playing music using the Mopidy-Youtube plugin again and it works well, including auto-play and its all nicely displayed and controllable within Mopidy-Iris. There are a lot of configuration and installation options, though. In my experience when it failed playing in the past it was mostly because of one of the following reasons:
* not using [yt-dlp](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/conf/mopidy-youtube.conf#L4) but the old youtube-dlc library that is not maintained anymore.
* a breaking change within Youtube that required me to update the yt-dlp library (which the community usually updates within days if not hours in case of a breaking Youtube change).
* the plugin supports two Youtube integration modes (Youtube Music or Youtube Data API) and you're using the one that is currently broken (due to breaking change in Youtube or so) - I just successfully tested it using the Youtube Data API integration which also lets you listen to songs from regular Youtube, not just Youtube Music.
* at some point it didn't play music when you didn't configure its authentication options (Youtube Data API key, Youtube Music cookie/auth headers) but last time I tried playing unauthenticated it worked again but that may have changed again - currently I have it configured with [authentication](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/deploy/app/configschema.yaml#L23-L54).

Fwiw my Mopidy installation including Mopidy-Youtube and Iris is containerized and can be found [here](https://github.com/mgoltzsche/mopidy-container). The compose file might be a bit outdated, though, since I am running it on Kubernetes/[Kubemate](https://github.com/mgoltzsche/kubemate) - the latter also provides a generated configuration GUI for the container's env vars based on the definition [here](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/deploy/app/configschema.yaml). You could use the docker container (`ghcr.io/mgoltzsche/mopidy:v0.18.0`) as is or compare the [Python package versions](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/Dockerfile#L19-L40) and [configuration](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/conf/mopidy-youtube.conf) I am using with your setup (though, my container [generates](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/entrypoint.sh#L106-L119) part of the configuration at runtime based on environment variables that are passed to it, particularly for [authentication](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/deploy/app/configschema.yaml#L23-L54)).

UPDATE: I realize I am not using the latest Mopidy-Youtube release but the [latest commit on its develop branch](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/Dockerfile#L39-L40), in case that matters.

**Comment by cma84** [comment]
> [@cma84](https://github.com/cma84) I just tried playing music using the Mopidy-Youtube plugin again and it works well, including auto-play and its all nicely displayed and controllable within Mopidy-Iris. There are a lot of configuration and installation options, though. In my experience when it failed playing in the past it was mostly because of one of the following reasons:
> 
>     * not using [yt-dlp](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/conf/mopidy-youtube.conf#L4) but the old youtube-dlc library that is not maintained anymore.
> 
>     * a breaking change within Youtube that required me to update the yt-dlp library (which the community usually updates within days if not hours in case of a breaking Youtube change).
> 
>     * the plugin supports two Youtube integration modes (Youtube Music or Youtube Data API) and you're using the one that is currently broken (due to breaking change in Youtube or so) - I just successfully tested it using the Youtube Data API integration which also lets you listen to songs from regular Youtube, not just Youtube Music.
> 
>     * at some point it didn't play music when you didn't configure its authentication options (Youtube Data API key, Youtube Music cookie/auth headers) but last time I tried playing unauthenticated it worked again but that may have changed again - currently I have it configured with [authentication](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/deploy/app/configschema.yaml#L23-L54).
> 
> 
> Fwiw my Mopidy installation including Mopidy-Youtube and Iris is containerized and can be found [here](https://github.com/mgoltzsche/mopidy-container). The compose file might be a bit outdated, though, since I am running it on Kubernetes/[Kubemate](https://github.com/mgoltzsche/kubemate) - the latter also provides a generated configuration GUI for the container's env vars based on the definition [here](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/deploy/app/configschema.yaml). You could use the docker container (`ghcr.io/mgoltzsche/mopidy:v0.18.0`) as is or compare the [Python package versions](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/Dockerfile#L19-L40) and [configuration](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/conf/mopidy-youtube.conf) I am using with your setup (though, my container [generates](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/entrypoint.sh#L106-L119) part of the configuration at runtime based on environment variables that are passed to it, particularly for [authentication](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/deploy/app/configschema.yaml#L23-L54)).
> 
> UPDATE: I realize I am not using the latest Mopidy-Youtube release but the [latest commit on its develop branch](https://github.com/mgoltzsche/mopidy-container/blob/v0.18.0/Dockerfile#L39-L40), in case that matters.

I'm still desperate, there's no way, I can't play the Youtube content.

I have uninstalled the whole plugin and reinstalled it, made sure it is using yt-dlp but then nothing.

Here is the log:


```
Mar 14 11:30:28 raspberrypi mopidy[2323021]: ERROR    [YouTubeCoreListener-11 (_actor_loop)] mopidy.listener Triggering event failed: tracklist_changed()
Mar 14 11:30:28 raspberrypi mopidy[2323021]: Traceback (most recent call last):
Mar 14 11:30:28 raspberrypi mopidy[2323021]:   File "/usr/lib/python3/dist-packages/mopidy/listener.py", line 44, in on_event
Mar 14 11:30:28 raspberrypi mopidy[2323021]:     getattr(self, event)(**kwargs)
Mar 14 11:30:28 raspberrypi mopidy[2323021]:   File "/usr/local/lib/python3.11/dist-packages/mopidy_youtube/backend.py", line 61, in tracklist_changed
Mar 14 11:30:28 raspberrypi mopidy[2323021]:     [video.audio_url for video in videos]
Mar 14 11:30:28 raspberrypi mopidy[2323021]:   File "/usr/local/lib/python3.11/dist-packages/mopidy_youtube/backend.py", line 61, in <listcomp>
Mar 14 11:30:28 raspberrypi mopidy[2323021]:     [video.audio_url for video in videos]
Mar 14 11:30:28 raspberrypi mopidy[2323021]:      ^^^^^^^^^^^^^^^
Mar 14 11:30:28 raspberrypi mopidy[2323021]:   File "/usr/local/lib/python3.11/dist-packages/mopidy_youtube/youtube.py", line 38, in wrapper
Mar 14 11:30:28 raspberrypi mopidy[2323021]:     func(self)  # should create the future
Mar 14 11:30:28 raspberrypi mopidy[2323021]:     ^^^^^^^^^^
Mar 14 11:30:28 raspberrypi mopidy[2323021]:   File "/usr/local/lib/python3.11/dist-packages/mopidy_youtube/youtube.py", line 390, in audio_url
Mar 14 11:30:28 raspberrypi mopidy[2323021]:     youtube_dl = importlib.import_module(youtube_dl_package)
Mar 14 11:30:28 raspberrypi mopidy[2323021]:                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 14 11:30:28 raspberrypi mopidy[2323021]:   File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
Mar 14 11:30:28 raspberrypi mopidy[2323021]:     return _bootstrap._gcd_import(name[level:], package, level)
Mar 14 11:30:28 raspberrypi mopidy[2323021]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Mar 14 11:30:28 raspberrypi mopidy[2323021]:   File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
Mar 14 11:30:28 raspberrypi mopidy[2323021]:   File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
Mar 14 11:30:28 raspberrypi mopidy[2323021]:   File "<frozen importlib._bootstrap>", line 1142, in _find_and_load_unlocked
Mar 14 11:30:28 raspberrypi mopidy[2323021]: ModuleNotFoundError: No module named 'yt-dlp'
Mar 14 11:30:28 raspberrypi mopidy[2323021]: ERROR    [YouTubeBackend-3 (_actor_loop)] mopidy_youtube translate_uri error "No module named 'yt-dlp'"
Mar 14 11:30:28 raspberrypi mopidy[2323021]: WARNING  [Core-8 (_actor_loop)] mopidy.core.tracklist Track is not playable: youtube:video:_dmOgDlWAkU

```

Would the container you indicate be valid for docker? 
At the time I tried to deploy it by docker but I did not find an updated development and finally I opted to install it as a service on the Raspberry, but what a headache it is giving me.

At the moment I am not using the API key configuration.

Thank you very much for the help you are offering me.

**Comment by mgoltzsche** [comment]
> `ModuleNotFoundError: No module named 'yt-dlp'`

That's the culprit. You didn't install `yt-dlp`.

> Would the container you indicate be valid for docker?

Yes, the container can be run using docker.

If you have another question regarding the mopidy-youtube plugin, I suggest to create an issue within the [mopidy-youtube](https://github.com/natumbri/mopidy-youtube) repo and link it here so that we're not polluting this issue with our actually unrelated conversation.
